src/core/fetchers.ts

```diff
@@ -0,0 +1,151 @@
+import type { Asset, ContentfulClientApi, Entry } from 'contentful'
+import { isExperienceEntry } from '../typeguards'
+
+type fetchExperienceEntitiesArgs = {
+  client: ContentfulClientApi<undefined>
+  experienceEntry: Entry
+  locale: string
+}
+
+export const fetchExperienceEntities = async ({
+  client,
+  experienceEntry,
+  locale,
+}: fetchExperienceEntitiesArgs) => {
+  if (!client) {
+    throw new Error(
+      'Failed to fetch experience entities. Required "client" parameter was not provided'
+    )
+  }
+
+  if (!locale) {
+    throw new Error(
+      'Failed to fetch experience entities. Required "locale" parameter was not provided'
+    )
+  }
+
+  if (!isExperienceEntry(experienceEntry)) {
+    throw new Error(
+      'Failed to fetch experience entities. Provided "experienceEntry" does not match experience entry schema'
+    )
+  }
+
+  const entryIds: string[] = []
+  const assetIds: string[] = []
+
+  for (const dataBinding of Object.values(experienceEntry.fields.dataSource)) {
+    if (!('sys' in dataBinding)) {
+      continue
+    }
+    if (dataBinding.sys.linkType === 'Entry') {
+      entryIds.push(dataBinding.sys.id)
+    }
+    if (dataBinding.sys.linkType === 'Asset') {
+      assetIds.push(dataBinding.sys.id)
+    }
+  }
+
+  const [entriesResponse, assetsResponse] = await Promise.all([
+    entryIds.length > 0 ? client.getEntries({ 'sys.id[in]': entryIds, locale }) : { items: [] },
+    assetIds.length > 0 ? client.getAssets({ 'sys.id[in]': assetIds, locale }) : { items: [] },
+  ])
+
+  return {
+    entries: (entriesResponse.items ?? []) as Entry[],
+    assets: (assetsResponse.items ?? []) as Asset[],
+  }
+}
+
+type fetchExperienceEntryArgs = {
+  client: ContentfulClientApi<undefined>
+  experienceTypeId: string
+  locale: string
+  idenifier: {
+    slug?: string
+    id?: string
+  }
```

Comment:
```suggestion
  idenifier: {
    slug: string
  } | {
    id: string
  }
```
I guess there is no use case where we want to allow the user to pass an empty object without any identifier.

Reply:
I don't think it works

<img width="663" alt="Screenshot 2023-10-23 at 14 52 53" src="https://github.com/contentful/experience-builder/assets/4171202/cdb4ae1d-009c-4b0a-8683-7f02bef5fdcf">